### ISSUE TITLE
Context-specific alt text

### ISSUE THREAD

**Issue by zerocrates** [omeka/omeka-s#1711]
(copied from zerocrates/AltText#3 by @triplingual, to account for this functionality being merged into the core)

> What would it take to incorporate this module's editing functionality in blocks, to allow for context-specific alt text? The accessibility pros at my uni say that good accessibility practice is for alt text of an image used as link content to declare the link target.
> 
> Doing that would mean needing to set alt text differently in a Media Embed block (where a still image links to the Item page) from the Item page (where the image links to its Media page) as well as from the Media page (where our images link to their files). I haven't set foot in the world of module code, so I don't know how complicated this would be.
> 
> it doesn't look like this has been discussed on the forums, but I'm happy to take the conversation there if it's more appropriate.

A followup comment:

> Totally. This is prompted partly by wanting to write locally specific doco for our users and then strengthened by talking with our web accessibility team. That team tells me that the alt for an image used as a link (which will be all uses of Media on a Site Page, yes?) should be the target for the link so adaptive tech compiling the links will have a readable and useful list. This target is settable in Site settings.
> 
> I have no idea how it would be done and how nontrivial it would be, but there's a place (+/- what you mention) to put a caption on a piece of media when adding it as an attachment to a block. That seems like the best (UX-wise) place to put the ability to give that attachment locally specific alt text. In the world of rainbows and unicorns, an alt text field in that attachment selection drawer would default to the alt text added at the Media level so that if it's fine the author doesn't need to do anything.
> 
> We've arrived at telling our people to use "See more about [appropriate media/item title]" as the alt text because that phrasing is mostly semantically valid for the 3 possible uses of Media I know of.

**Comment by zerocrates** [comment]
@triplingual: one technically-simple option here is to allow (as an option) the Caption text to be used as the alt instead of being a visible caption.

Does this make sense at all, or do you really want to have it as something separate?

**Comment by triplingual** [comment]
Thanks for pinging me, and sorry for the delay. As an intermediate step, this can have some value. The way you're looking at it, would the alt text globally assigned to the Media be used only if the user does not explicitly say to use the Caption text?

The approach makes sense in some situations for assigning context-specific alt text for a non-visual browser user, but sounds like it then deprives the sighted user of the context-specific text. If you have alt text and a Caption, good practice is to have them different, with the Caption holding some amount more of description. The approach you're describing means a Page creator is choosing among global alt text and a Caption (where the user has chosen to not override the global alt text with the Caption text), global alt text and no Caption (ditto), or custom alt text and no Caption (user has chosen to override).

Those options can work, but there is still a difficulty in having global alt text that works in contexts across user-customized alt text and globally assigned alt text. It's hard to write alt text to be used in the above situations and to be used where there is Media added without the ability to customize the alt text (Item page, Item with Metadata Block, browse Items, or browse Item Set[s]). Where it's not possible to customize the alt text, the user could have something like "Click to read more about [Item title]" as the global alt text, but in a Media Embed block where the user wants a descriptive Caption for visual users, that alt text is sub-optimal.

We could tell our users to make the alt text "{short desc} More information adjacent.", though that puts a burden on the person using a non-visual browser to figure out where the adjacency is. Is it in adjacent markup? A Caption? At a link target? If the alt text were just a short description (to accommodate the Item with Metadata and Browse Item Set cases), the user could still have the caption be a longer description, but the non-visual browser user then isn't informed of the connection between the alt and the Caption text.

Some of this comment is me trying to put my thinking out there so you see why I think this is a good idea, but still leaves room for issues to crop up in our use of Omeka.

**Comment by triplingual** [comment]
It's not so much the name as something that explains it for a non-visual browser to read out, as I understand it, though that may often be a name. The rule of thumb our web accessibility folks give out is that the alt text should "convey the purpose & function of the image." Media that only illustrate text on page could be considered "decorative" and not need alt text at all. An image can also be complex enough that it should get alt text that points the visitor to another element for longer description or discussion.

Ignoring development constraints and difficulty, yes, having a Caption box and an Alt Text box everywhere Media get inserted would be more nearly ideal. It would probably be an additional good step to populate that Alt Text box with the global alt text. Then it's an issue of us making sure our curators know they should think about the alt text to be sure it's an accurate description of the purpose and function of the image in the Page where they are putting it, and even to blank it out if the image is just decorative. Blocks or other widgets that don't take customization of Media elements (Item with Metadata, browse Items, browse Item Set) could still get global alt text, but the purpose and function of the Media thumbnails in those three cases seem similar enough to me that it would be easier to write the global alt text to apply to them.

(Apologies if this is all repetitive accessibility talk for you.)
